src/mongocxx/collection.cpp

```diff
@@ -1149,10 +1149,18 @@ cursor collection::_distinct(const client_session* session,
         throw bsoncxx::exception{bsoncxx::error_code::k_internal_error};
     }
 
+    const bson_t* error_document;
```

Comment:
TLDR; Here, you could use a `scoped_bson_t` instead of a pointer to a `bson_t`.

Because the lifetime of objects is less obvious in C++ than C we opt to use classes that manage their own life rather than pointers, which we later have to remember to clean up.  

For example, if an exception is thrown between a heap-allocated object's creation and a call to some destructor function then the destructor function will never get called and we'll leak memory. C functions don't throw exceptions but most C++ functions do.

Reply:
Sounds good! Thanks for the explanation here and offline. `cursor_error_document` does require a `**bson_t`, so I used `bson_for_init` to make another `error_doc` variable first; let me know if there's a better way.